Added option to filter tests

SebastianZ opened the pull request:
This addresses #178 and adds a drop-down list to the sidebar allowing to filter the tests by specification status. The options are "All", "Stable", and the different CSS snapshots. The chosen option is saved in the local storage and read on page load to keep the shown info stable.

For that to work I added a new `status` field to the specifications which holds the field `stability` indicating whether a spec. is `stable` or `experimental`. Optionally, there are two fields `first-snapshot` and `last-snapshot` indicating the first and last snapshot a specification got included as official part of CSS, respectively. `last-snapshot` indicates that a higher level spec. supersedes a previous one in a later snapshot. I.e. it ensures that only the level listed in the snapshot is shown.

Note that this is still WIP. In its current form I had to drop the counting up animation. Instead, the results are shown right away. Also, I'd like to get some feedback on whether that's generally the right direction I am heading. Are you missing any options? Is the idea of superseding specs ok? Should the default option be changed? Should the options be more fine-grained (maybe even be feature based as discussed earlier)? Should specific features listed in "Safe to Release pre-CR Exceptions" be included even when their specification isn't?

Sebastian

Zefling commented:
It's really cool is much faster. Anchors work better.

There is something weird:
![image](https://user-images.githubusercontent.com/3032555/148259190-cf7cc5be-6238-46ea-bd87-6222df2f739d.png)

SebastianZ commented:
> It's really cool is much faster. Anchors work better.

Yes, that's a benefit.

> There is something weird: ![image](https://user-images.githubusercontent.com/3032555/148259190-cf7cc5be-6238-46ea-bd87-6222df2f739d.png)

It was reset for every test. I've fixed that now.

> And it would be nicer if we could filter those in "experimental".

Good point! I've added a filter for them.

Sebastian

LeaVerou commented:
I did quite like the animation, and it allows us to have slower tests too, but I'll concede if you both agree with removing it. I'm not one to hold a project back just because of my opinions. :) 

Is there a screenshot or something about this change? It would help review it!

LeaVerou commented:
> Note that this is still WIP.

You may want to mark it as draft until it's ready to merge

SebastianZ commented:
I also liked the animation somewhat, though it also had some issues. Firstly, the mentioned anchor problem, i.e. you can't get to the spec. by entering a URL with its anchor. Also, with more tests it takes longer and longer until the final values appear. And the displayed time it took to execute the tests is a bit misleading as it takes much longer until they are displayed, so users might not immediately understand what that value means.

So, at least for now, I am for removing it. We may add back an animation solely for the display of the score at some point.

I think with the latest commits the changes aren't WIP anymore. Again, detailed feedback is very welcome.

Here's a screenshot:

![Screenshot of patch showing filter drop-down list](https://user-images.githubusercontent.com/958943/148338757-beb6c158-6300-4b47-9b00-b291ce7e6b9b.png)

Sebastian

Zefling commented:
The animation was a cool at first, but now it becomes a problem, because there is really a lot of testing. Even to test the addition of new tests, it is little useless, because you have to wait longer and longer. I think it would be beneficial for everyone to remove it.

SebastianZ commented:
> > And it would be nicer if we could filter those in "experimental".
> 
> Good point! I've added a filter for them.

Forgot to push that change. Did that now.

Sebastian

Zefling commented on csstest.js:
```diff
@@ -431,6 +431,8 @@ function runTests(filter = '') {
 		// Filter list of specifications
 		if (filter === 'stable' && Specs[spec]?.status?.stability !== 'stable') {
 			continue;
+		} else if (filter === 'experimental' && Specs[spec]?.status?.stability === 'stable') {
```
This syntax not works on older browsers.

SebastianZ replied:
Optional chaining is already [supported in more than 90% of browsers according to caniuse](https://caniuse.com/mdn-javascript_operators_optional_chaining). So that choice was deliberate. Though I can replace it by old syntax.

SebastianZ replied:
I've removed the optional chaining now.

LeaVerou commented:
Now that I saw the screenshot I understand what the PR is about. What a fantastic feature!!! Thank you so much for implementing it!

SebastianZ commented:
> Maybe we should move "whatwg" and "svg" in separate group.

Let's discuss this in a separate issue, please! I'd like to get this feature online first. Afterwards we can still iterate on it and add more filters.

I'll also create new features for the questions I had above, so we can discuss them separately and iterate on what we have now.

Sebastian

PolariTOON commented on csstest.js:
```diff
@@ -401,6 +404,83 @@ function passclass(info) {
+}
+
+function runTests(filter = '') {
+	let specs = [];
```
I see you replaced some `var` keywords by `let`.
Do we want to embrace ES6-style javascript now? or still being somehow compatible with old browsers?

Zefling replied:
IE isn't really dead yet. For me, it would be necessary to wait until it's definitely dead (removal of Windows). I let you decide.

SebastianZ replied:
According to [caniuse.com](https://caniuse.com/let), IE has (partial) support for `let`. The biggest percentage of a non-supporting browser is Opera Mini with ~1%. And the overall support is at ~95%, which seems pretty stable to me.

Personally, I prefer to use ES6+ features as long as there is good browser support for them. If you two believe we should strive for full backwards compatibility, I could change it back to `var`.
Though I have to admit, I never tried to call the site in a very old browser so far. I assume there are other pieces of code that might be unsupported in older browsers.

Having said that, I am also planning to push #57 forward which requires a good browser coverage. So maybe that requires the JS part to be backwards compatible.

Sebastian

SebastianZ replied:
> For me, it would be necessary to wait until it's definitely dead (removal of Windows).

For what it's worth, IE is removed in Windows 11 and finally [will be gone in June this year](https://www.microsoft.com/en-us/download/internet-explorer.aspx), though it will surely still take many years until nobody uses IE anymore.
According to [statcounter](https://gs.statcounter.com/browser-market-share) IE still has a market share of 0.46%.

Sebastian

SebastianZ replied:
Because there's a mixture of `var` and `let` after that patch, I decided to roll back to `var` again for now.

I'll file a follow-up issue on refactoring the code base, instead.

Sebastian

SebastianZ replied:
I've filed #216 now to discuss whether and which ES6+ features we want to use. Feel free to comment there!

Sebastian

PolariTOON replied:
Could this line also be reverted to `var`?

SebastianZ replied:
Done. Thank's for the hint!

PolariTOON commented on csstest.js:
```diff
@@ -401,6 +404,83 @@ function passclass(info) {
+
+	for (let spec in Specs) {
+		// Filter list of specifications
+		if (filter === 'stable' && Specs[spec]?.status && Specs[spec].status.stability !== 'stable') {
```
There is still one occurrence of the "?." operator here.
I have no strong opinion against that if we want to only target really modern browsers now.

SebastianZ replied:
Oh, sorry! Removed it now.

Sebastian

PolariTOON commented on filter.js:
```diff
@@ -0,0 +1,7 @@
+(function() {
+    ele('filter').addEventListener('change', evt => {
```
I'm feeling a bit sad that all the tests have to be reexecuted each time we select a new filter.
The UI becomes unresponsive for half a second when selecting "All".

SebastianZ replied:
I'll see if I can get it to just execute once.

Sebastian

SebastianZ replied:
I realize now that its more refactoring than I anticipated for this patch. The calculation and output logic are currently strongly entangled.

Let's split that out into a refactoring patch.

Sebastian

PolariTOON replied:
Ok let's leave that for a follow-up PR

PolariTOON reviewed (commented):
Overall, LGTM, i think this is a useful feature.
This allows to see how both the CSS language, and the support for it evolve.

But in practice, i don't think this reduces the noise that much : there's a lot of green in the snapshots coming from the previous better supported versions.
We could still add diffs between snapshots later, like in Kangax's compatibility table for JavaScript.

Just one more question: what makes a spec "stable" or "experimental"?

SebastianZ commented:
@PolariTOON Thank you for the detailed feedback!

> But in practice, i don't think this reduces the noise that much : there's a lot of green in the snapshots coming from the previous better supported versions. We could still add diffs between snapshots later, like in Kangax's compatibility table for JavaScript.

That's true. Currently, the snapshot filters include all specifications listed as part of the official CSS definition of the related snapshot. Adding diffs between two snapshots shouldn't be too hard.

> Just one more question: what makes a spec "stable" or "experimental"?

Sorry, I added a description to the filters in their `title` attributes but didn't make that clear in the description!

I defined "stable" as everything that's listed in the latest snapshot plus all the predecessor specifications and those that probably won't change much anymore.
As described in #178, "stable" could also be defined as something that's implemented in at least two different browsers, though that may require a property-based approach.

Sebastian

SebastianZ commented:
@PolariTOON I think I've addressed all your points now. For most of them I've created separate issues to discuss them further.

If there's anything left you think should be in this first version of the feature, please tell me! Otherwise, it would be great if you also gave your approval, so it can be merged.

Sebastian

PolariTOON commented on csstest.js:
```diff
@@ -394,6 +394,73 @@ function passclass(info) {
+
+function runTests(filter = '') {
+	let specs = [];
+	let timeBefore = +new Date;
```
Also this one?

SebastianZ replied:
Done.

PolariTOON reviewed (approved):
Just two minor nits to address and then i think this is ready to land as a first version.
We can refine the feature and refactor later, indeed.
